OpenCL: POCL fixes

magnumripper opened the pull request:
This stops GPG and DEScrypt OpenCL formats from crashing under POCL (as tested on Ubuntu 20.04) - making all formats pass self-test. It also adds a quick bail-out for when LWS auto-tune takes too long.

magnumripper committed:
OpenCL: POCL fixes

This stops GPG and DEScrypt OpenCL formats from crashing under POCL
(as tested on Ubuntu 20.04) - making all formats pass self-test.

Closes #4684

magnumripper commented on src/opencl_common.c:
```diff
@@ -1867,6 +1867,8 @@ void opencl_find_best_lws(size_t group_size_limit, int sequential_id,
 			fprintf(stderr, " %ss%s\n", ns2string(sumRunTime),
 			    ((double)(sumRunTime) / kernelExecTimeNs < 0.997)
 			        ? "+" : "");
+		if (sumRunTime > 2 * kernelExecTimeNs)
+			break;
```
This is for the issue [reported here](https://github.com/openwall/john/issues/4684#issuecomment-824831938), which was intermittent. I figured if we *ever* get over twice as long duration from doubling LWS but with same GWS, we can't be on the way to anything good, right?

solardiz replied:
Doubled duration with same GWS means twice lower speeds, so, yes, I guess it means the doubled LWS wasn't a move in the right direction. I wonder, though, whether exceptions exist with LWS/GWS values way out of the device's optimal range, where further GWS adjustment would bring them into the optimal range. Hopefully not.

Maybe this should be a separate commit with its own description.

solardiz commented on run/opencl/opencl_DES_kernel_params.h:
```diff
@@ -11,10 +11,10 @@ typedef unsigned WORD vtype;
+ * same goes for MESA and POCL.
  */
-#if nvidia_sm_5x(DEVICE_INFO) || gpu_intel(DEVICE_INFO) || __MESA__ ||  \
+#if nvidia_sm_5x(DEVICE_INFO) || __OS_X__ || __MESA__ || __POCL__ || \
```
Need to test how this affects Intel GPUs under non-macOS, or/and don't drop `gpu_intel(DEVICE_INFO)` in the same PR/commit. Can you test on "well"?

magnumripper replied:
Done now, There's virtually no difference in speed.

solardiz commented on doc/README-OPENCL:
```diff
@@ -17,8 +17,9 @@ distribution, only the vendor-supplied drivers support OpenCL.  Either
-be usable in some OpenCL formats.
+Notice: Beignet and Mesa are not officially supported, but may be usable
+with some OpenCL formats.  POCL on the other hand has now been seen passing
+all self-tests (tested on Ubuntu 20.04).
```
I'd add: "tested on Ubuntu 20.04 with bleeding-jumbo as of 2021/04/22".

solardiz reviewed (approved):
Looks good to me overall, but I made some comments.

magnumripper committed:
OpenCL LWS auto-tune: Bail out if new size was way slower

If speed was worse than half, we don't expect even larger LWS to be
any better.  This was added as a workaround for a POCL problem where
LWS=8 would be fine at 400 ms but all larger sizes taking 11 seconds
or more each, leading to an extremely slow auto-tune for no benfit.

See #4684

magnumripper committed:
OpenCL: Allow --force-vector and --force-scalar along with --test

This was inadvertently disallowed with the recent options review.

magnumripper commented on src/options.c:
```diff
@@ -63,7 +63,8 @@ static char *costs_str;
 /* Common req_clr for --test, --test-full and --stress-test */
 #define TEST_REQ_CLR (~FLG_TEST_SET & ~FLG_FORMAT & ~FLG_SAVEMEM & ~FLG_MASK_CHK & ~FLG_NO_MASK_BENCH & \
-                      ~FLG_VERBOSITY & ~FLG_INPUT_ENC & ~FLG_SECOND_ENC & ~GETOPT_FLAGS & ~FLG_NOTESTS)
+                      ~FLG_VERBOSITY & ~FLG_INPUT_ENC & ~FLG_SECOND_ENC & ~GETOPT_FLAGS & ~FLG_NOTESTS & \
+                      ~FLG_SCALAR & ~FLG_VECTOR)
```
This added as well now (separate commit).
